lib/kubernetes-deploy/restart_task.rb

```diff
@@ -27,58 +23,71 @@ def initialize(context:, namespace:, logger:)
       @context = context
       @namespace = namespace
       @logger = logger
-      @kubeclient = build_v1_kubeclient(context)
-      @v1beta1_kubeclient = build_v1beta1_kubeclient(context)
-      @policy_v1beta1_kubeclient = build_policy_v1beta1_kubeclient(context)
     end
 
     def perform(deployments_names = nil)
+      start = Time.now.utc
       @logger.reset
+
+      @logger.phase_heading("Initializing restart")
       verify_namespace
+      deployments = identify_target_deployments(deployments_names)
+
+      @logger.phase_heading("Triggering restart by touching ENV[RESTARTED_AT]")
+      patch_kubeclient_deployments(deployments)
+
+      @logger.phase_heading("Waiting for rollout")
+      resources = build_watchables(deployments, start)
+      ResourceWatcher.new(resources, logger: @logger, operation_name: "restart").run
+      success = resources.all?(&:deploy_succeeded?)
+    rescue FatalDeploymentError => error
+      @logger.summary.add_action(error.message)
+      success = false
```

Comment:
The flow is what variable is returned is a bit opaque.
Do you think this version would be a bit better readable?

```ruby
def perform
  # do things that might raise
  statsd_report(start, status: "success")
  true
rescue =>
  statsd_report(start, status: "failed")
  false
ensure
  @logger.print_summary
end
```

Reply:
That won't work as is because `print_summary` needs to know the success status, and the condition that didn't raise might still have failed (that's why it does `success = resources.all?(&:deploy_succeeded?)`). So that'd end up as:

```ruby
      success = resources.all?(&:deploy_succeeded?)
      report_statsd(start, success: success, num_deployments: deployments.to_a.length)
      success
    rescue FatalDeploymentError => error
      @logger.summary.add_action(error.message)
      report_statsd(start, success: false, num_deployments: deployments.to_a.length)
      success = false
    ensure
      @logger.print_summary(success)
    end
```

Which seems pretty much the same to me, except that the statsD no longer leverages `status`. What do you find unclear about it? I note that the return value of this is now heavily tested.